docs/proposals/ns-scoped-catalog.md

```diff
@@ -0,0 +1,256 @@
+```go
+type ServiceBrokerSpec struct {
+    SharedServiceBrokerSpec `json:",inline"`
+    
+    AuthInfo *ServiceBrokerAuthInfo `json:"authInfo,omitempty"`
+}
+
+type ServiceBrokerStatus struct {
+    SharedServiceBrokerStatus `json:",inline"`
+}
+
+type ServiceClassSpec struct {
+    SharedServiceClassSpec `json:",inline"`
+    
+    ServiceBrokerName string `json:"serviceBrokerName"`
+}
+
+type ServiceClassStatus struct {
+    SharedServiceClassStatus `json:",inline"`
+}
+
+type ServicePlanSpec struct {
+    SharedServicePlanSpec `json:",inline"`
+    
+    ServiceBrokerName `json:"serviceBrokerName"`
+    
+    ServiceClassRef LocalObjectReference `json:"serviceClassRef"`
+}
+
+type ServicePlanStatus struct {
+    SharedServicePlanStatus `json:",inline"`
+}
+```
+
+After this step, we will have the namespaced resources for brokers, services, and plans, but they will not be functional yet.
+
+### Add control loops for ns-scoped resources
+
+Next, we will add control loops for the new namespaced resources for brokers, services, and plans.
+
+We'll also add associated tests (similar to broker integration tests that already exist).
+
+After this step, users will be able to add a broker to a namespace and see the namespaced services and plans populated, but they won't be able to provision an instance of a namespaced service/plan yet.
+
+### Make it possible to provision instances of ns-scoped services and plans
+
+Next, we will make it possible to provision an instance of a namespaced service and plan. This involves:
+
+- Adding new fields to `ServiceInstanceSpec` for users to indicate which ns-scoped services they want
+- Adding new fields to `ServiceInstanceSpec` for references to ns-scoped resources
+- Associated validations / fuzzers for those fields
+- Modifying the reference subresource to set the new reference fields appropriately
+- Modifying the controller to resolve the new specification fields and set reference fields
+
+Relevant API resources at this step:
+
+```go
+type PlanReference struct {
+    // existing fields omitted
+    
+    ServiceClassExternalName string `json:"serviceClassExternalName"`
+    ServiceClassName         string `json:"serviceClassName"`
+    ServicePlanExternalName  string `json:"servicePlanExternalName"`
+    ServicePlanName          string `json:"servicePlanName"`
+}
+
+type ServiceInstanceSpec struct {
+    // existing fields omitted
+
+    ServiceClassRef *LocalObjectReference `json:"serviceClassRef,omitempty"`
+    ServicePlanRef  *LocalObjectReference `json:"servicePlanRef,omitempty"`
+}
+```
+
+After this step, users will be able to provision (but not bind to) ns-scoped services and plans.
+
+### Make it possible to bind to instances of ns-scoped services and plans
+
+Next, we will make the required changes to the binding controller loop to make it possible to create bindings against service instances that are associated with the ns-scoped variants of serviceclasses and plans.
+
+### Remove feature flag
```

Comment:
This is the first time feature flag is mentioned in this proposal, i.e. there is no "add feature flag" anywhere.
Also, I'm not even sure we need feature flag at all, given that the proposed changes are backward compatible. Is that to be able to iteratively implement this proposal, i.e. having it merged without enabling for a first couple iterations until we get MVP? I would probably prefer to have an MVP in the first PR.

Reply:
I think he means feature gate

Reply:
Yep, feature gate.